src/timeline/NhekoDBusInterface.cpp

```diff
@@ -0,0 +1,168 @@
+                           const QString &title,
+                           const QImage &image,
+                           QObject *parent)
+  : QObject{parent}
+  , roomId_{mxid}
+  , alias_{alias}
+  , roomName_{title}
+  , image_{image}
+{}
+
+RoomInfoItem::RoomInfoItem(const RoomInfoItem &other)
+  : QObject{other.parent()}
+  , roomId_{other.roomId_}
+  , alias_{other.alias_}
+  , roomName_{other.roomName_}
+  , image_{other.image_}
+{}
+
+void
+RoomInfoItem::init()
+{
+    qDBusRegisterMetaType<RoomInfoItem>();
+    qDBusRegisterMetaType<QVector<RoomInfoItem>>();
+    qDBusRegisterMetaType<QImage>();
+}
+
+RoomInfoItem &
+RoomInfoItem::operator=(const RoomInfoItem &other)
+{
+    roomId_   = other.roomId_;
+    alias_    = other.alias_;
+    roomName_ = other.roomName_;
+    image_    = other.image_;
+    return *this;
+}
+
+QDBusArgument &
+operator<<(QDBusArgument &arg, const RoomInfoItem &item)
+{
+    arg.beginStructure();
+    arg << item.roomId_ << item.alias_ << item.roomName_ << item.image_;
+    arg.endStructure();
+    return arg;
+}
+
+const QDBusArgument &
+operator>>(const QDBusArgument &arg, RoomInfoItem &item)
+{
+    arg.beginStructure();
+    arg >> item.roomId_ >> item.alias_ >> item.roomName_ >> item.image_;
+    if (item.image_.isNull())
+        item.image_ = QIcon::fromTheme(QStringLiteral("group")).pixmap(32, 32).toImage();
+
+    arg.endStructure();
+    return arg;
+}
+
+/**
+ * Automatic marshaling of a QImage for org.freedesktop.Notifications.Notify
+ *
+ * This function is heavily based on a function from the Clementine project (see
+ * http://www.clementine-player.org) and licensed under the GNU General Public
+ * License, version 3 or later.
+ *
+ * SPDX-FileCopyrightText: 2010 David Sansome <me@davidsansome.com>
+ */
+QDBusArgument &
+operator<<(QDBusArgument &arg, const QImage &image)
+{
+    if (image.isNull()) {
+        arg.beginStructure();
+        arg << 0 << 0 << 0 << false << 0 << 0 << QByteArray();
+        arg.endStructure();
+        return arg;
+    }
+
+    QImage scaled;
+    if (image.height() > 100)
+        scaled = image.scaledToHeight(100, Qt::SmoothTransformation);
+    else
+        scaled = image;
+    scaled = scaled.convertToFormat(QImage::Format_ARGB32);
+
+#if Q_BYTE_ORDER == Q_LITTLE_ENDIAN
+    // ABGR -> ARGB
+    QImage i = scaled.rgbSwapped();
+#else
+    // ABGR -> GBAR
+    QImage i(scaled.size(), scaled.format());
+    for (int y = 0; y < i.height(); ++y) {
+        QRgb *p   = (QRgb *)scaled.scanLine(y);
+        QRgb *q   = (QRgb *)i.scanLine(y);
+        QRgb *end = p + scaled.width();
+        while (p < end) {
+            *q = qRgba(qGreen(*p), qBlue(*p), qAlpha(*p), qRed(*p));
+            p++;
+            q++;
+        }
+    }
+#endif
+
+    arg.beginStructure();
+    arg << i.width();
+    arg << i.height();
+    arg << i.bytesPerLine();
+    arg << i.hasAlphaChannel();
+    int channels = i.isGrayscale() ? 1 : (i.hasAlphaChannel() ? 4 : 3);
```

Comment:
I thought I fixed this?

Reply:
I did in 9a2e07cbce9fab2eebc52923776522cb6303bae4

Reply:
Oh, I hadn't noticed your fix.

Reply:
The diff also doesn't show it, which confuses me

Reply:
I never pushed a rebase onto master (in progress).

Reply:
Rebased, but I should test first...